### PR TITLE
[To new_cluster] [IOTDB-2683] Hash algorithm for device group

### PR DESCRIPTION
**Pull request by CRZbulabula**
Hash algorithm test result: [General hash indexes test](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199531564)

**Changes:**

- The DeviceGroup hash algorithm is implemented as abstract class.
- Add ConfigNodeConf, ConfigNodeDescriptor and ConfigNodeConstant. To imitate the `.properties` file reading process of the single machine.

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 1]
```diff
@@ -30,20 +40,48 @@
...
+        hashExecutor = new BKDRHashExecutor(deviceGroupCount);
+        break;
+      case "AP":
+        hashExecutor = new APHashExecutor(deviceGroupCount);
+        break;
+      case "JS":
+        hashExecutor = new JSHash(deviceGroupCount);
+        break;
+      case "SDBM":
+        hashExecutor = new SDBMHash(deviceGroupCount);
```
Comment (wangchao316) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:79]: add case default, throw a exception("does not support Hash......")

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# Datatype: String
+# Currently, BKDR, AP, JS or SDBM hash algorithm is available
+# device_group_hash_algorithm=BKDR
+
+# DeviceGroupHashExecutor package name
+# Datatype: String
+# If you want to implement your own hash algorithm, you can inherit the DeviceGroupHashExecutor class and
+# modify this parameter to correspond to your Java package prefix.
+# But make sure that your own DeviceGroupHashExecutor class name is device_group_hash_algorithm + "HashExecutor".
+# device_group_hash_executor_package=org.apache.iotdb.confignode.manager.hash.
```
Comment (wangchao316) [inline confignode/src/assembly/resources/conf/iotdb-confignode.properties:38]: hi,  we could merge a config parameter. example:
device_group_hash_algorithm_class=org.apache.iotdb.confignode.manager.hash.BKDRHashExecutor.
2. remove device_group_hash_executor_package parameter

**Review by wangchao316 — CHANGES_REQUESTED** [review]

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties (hunk 2 above)
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties]: Good idea! I'm done.

**Review by wangchao316 — APPROVED** [review]
LGTM

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**